distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,969 @@
+        sockname = self._transport.get_extra_info("sockname")
+        if sockname is not None:
+            return unparse_host_port(*sockname[:2])
+        return "<unknown>"
+
+    @property
+    def peer_addr(self):
+        if self.is_closed:
+            return "<closed>"
+        peername = self._transport.get_extra_info("peername")
+        if peername is not None:
+            return unparse_host_port(*peername[:2])
+        return "<unknown>"
+
+    @property
+    def is_closed(self):
+        return self._transport is None
+
+    def _abort(self):
+        if not self.is_closed:
+            self._transport, transport = None, self._transport
+            transport.abort()
+
+    def _close_from_finalizer(self, comm_repr):
+        if not self.is_closed:
+            logger.warning(f"Closing dangling comm `{comm_repr}`")
+            try:
+                self._abort()
+            except RuntimeError:
+                # This happens if the event loop is already closed
+                pass
+
+    async def _close(self):
+        if not self.is_closed:
+            self._transport, transport = None, self._transport
+            transport.close()
+        await self._closed_waiter
+
+    def connection_made(self, transport):
+        # XXX: When using asyncio, the default builtin transport makes
+        # excessive copies when buffering. For the case of TCP on asyncio (no
+        # TLS) we patch around that with a wrapper class that handles the write
+        # side with minimal copying.
+        if type(transport) is asyncio.selector_events._SelectorSocketTransport:
+            transport = _ZeroCopyWriter(self, transport)
+        self._transport = transport
+        # Set the buffer limits to something more optimal for large data transfer.
+        self._transport.set_write_buffer_limits(high=512 * 1024)  # 512 KiB
+        if self.on_connection is not None:
+            self.on_connection(self)
+
+    def get_buffer(self, sizehint):
+        """Get a buffer to read into for this read event"""
+        # Read into the default buffer if there are no frames or the current
+        # frame is small. Otherwise read directly into the current frame.
+        if self._frames is None or self._frame_nbytes_needed < self._default_len:
+            self._using_default_buffer = True
+            return self._default_buffer[self._default_end :]
+        else:
+            self._using_default_buffer = False
+            frame = self._frames[self._frame_index]
+            return frame[-self._frame_nbytes_needed :]
+
+    def buffer_updated(self, nbytes):
+        if nbytes == 0:
+            return
+
+        if self._using_default_buffer:
+            self._default_end += nbytes
+            self._parse_default_buffer()
+        else:
+            self._frame_nbytes_needed -= nbytes
+            if not self._frames_check_remaining():
+                self._message_completed()
+
+    def _parse_default_buffer(self):
```

Comment:
General comment on this function, it seems we call a bunch of functions in here only once, which makes it a bit harder to read as one needs to hop around the code. Any thoughts on just inlining all of those calls here? Admittedly it would be longer with more logic, but it would be possible to read all the content in one place

Reply:
I like breaking them out into smaller simpler functions, it feels easier to me to read it as written. Each `_parse_*` method only has one task (parse the thing it's in charge of parsing, or return `False` if more data is needed). It also makes it easier to compose if we change our framing protocol later on.